File: aws-lambda/src/databricks_cdk/resources/permissions/cluster_policy_permissions.py
import logging
from typing import List, Union

from pydantic import BaseModel

from databricks_cdk.resources.permissions.models import GroupPermission, ServicePrincipalPermission, UserPermission
from databricks_cdk.utils import CnfResponse, put_request

logger = logging.getLogger(__name__)


class ClusterPolicyPermissionsProperties(BaseModel):
    action: str = "cluster-policy-permissions"
    workspace_url: str
    cluster_policy_id: str
    access_control_list: List[Union[UserPermission, GroupPermission, ServicePrincipalPermission]]


def get_cluster_policy_permissions_url(workspace_url: str, cluster_policy_id: str):
    """Getting url for creating policy permissions requests"""
    return f"{workspace_url}/api/2.0/permissions/cluster-policies/{cluster_policy_id}"


def create_or_update_cluster_policy_permissions(
    properties: ClusterPolicyPermissionsProperties,
) -> CnfResponse:
    """Create cluster policy permissions at databricks"""

    # Json data
    body = {
        "access_control_list": [a.dict() for a in properties.access_control_list],
    }
    put_request(
        f"{get_cluster_policy_permissions_url(properties.workspace_url, properties.cluster_policy_id)}",
        body=body,
    )
    return CnfResponse(
        physical_resource_id=properties.cluster_policy_id,
    )


def delete_cluster_policy_permissions(physical_resource_id: str) -> CnfResponse:
    """Deletes cluster policy permissions at databricks"""
    # no need to remove
    return CnfResponse(physical_resource_id=physical_resource_id)

File: typescript/src/resources/permissions/cluster-policy-permissions.ts
import {CustomResource} from "aws-cdk-lib";
import {Construct} from "constructs";
import {UserPermission, GroupPermission, ServicePrincipalPermission} from "./models";


export interface ClusterPolicyPermissionsProperties {
    workspaceUrl: string
    clusterPolicyId: string
    accessControlList: Array<UserPermission | GroupPermission | ServicePrincipalPermission>
}

export interface ClusterPolicyPermissionsProps extends ClusterPolicyPermissionsProperties {
    readonly serviceToken: string
}

export class ClusterPolicyPermissions extends CustomResource {
    constructor(scope: Construct, id: string, props: ClusterPolicyPermissionsProps) {
        super(scope, id, {
            serviceToken: props.serviceToken,
            properties: {
                action: "cluster-policy-permissions",
                workspace_url: props.workspaceUrl,
                cluster_policy_id: props.clusterPolicyId,
                access_control_list: props.accessControlList,
            }
        });
    }
}
